src/libraries/System.IO.Pipelines/tests/StreamPipeReaderTests.cs

```diff
@@ -31,18 +31,28 @@ public async Task CanRead()
             reader.Complete();
         }
 
-        [Fact]
-        public async Task CanReadAtLeast()
+        [Theory]
+        [InlineData(-1, false)]
+        [InlineData(-1, true)]
+        [InlineData(5, false)]
+        [InlineData(5, true)]
+        public async Task CanReadAtLeast(int bufferSize, bool bufferedRead)
```

Comment:
I think we need more tests. Maybe we can do a similar thing to what we did with the CopyToAsync tests? We need Pipe, StreamPipeReader and the base implementation (everything except the back pressure test).

Also, I would add a test that verifies you can't read more than max buffer chunks. Write  5MB and make sure multiple buffer segments are returned.

Reply:
I have grouped all common tests around ReadAtLeastAsync in one base test class, as for CopyToAsync tests.
I have also added a test to cover the MaxBufferSize logic.